### DISCUSSION TITLE
Any way to run a Tenant Template in Azure Functions without storing template in blob storage?

### DISCUSSION THREAD

**Discussion by prawnz** [pnp/powershell#469] [Q&A]
Is there any way currently to run Invoke-PnPTenantTemplate using Azure Functions without storing template in blob storage?

**Comment by BaronSparky** [comment]
**+1** You can _"Invoke-PnPSiteTemplate_" from a URL, it would be great to have the same feature for "_Invoke-PnPTenantTemplate_"

I want to store all tenant templates in a SPO list and use them from an Azure Function without having to copy them to the Azure function blob storage first....

**Reply by prawnz** [reply]
Same minds here.

**Comment by BaronSparky** [comment]
I was trying to use the `-CreateDrive` argument on `Connect-PnPOnline` with the bright idea of perhaps trying to `Invoke-PnPTenantTemplate -Path SPO:\sites\somesite\somelibrary\sometemplate.xml` but the fileprovider is always pre-appended so it looks for a template in "`C:\home\site\wwwroot\SPO:\sites\somesite\somelibrary\sometemplate.xml`"

`2021-04-20T14:09:58.235 [Error] ERROR: File C:\home\site\wwwroot\SPO:\sites\somesite\somelibrary\sometemplate.XML does not exist`

Tried `Set-Location` to the SPO: drive, etc. and various flavours of `"Invoke-PnPTenantTemplate -Path`" with no joy....

Any ideas as to how can we manage change in the templates that are stored in a SPO library???? @erwinvanhunen - Is "`-CreateDrive`" an option here and I am just doing something wrong??

If the template **HAS** to be in the function app root blob, then is there an easy way of copying the template file from SPO to the Azure function. I have tried using `Get-PnPFile` but get an access denied when trying to use that to copy to the Azure Function root directory.

**Comment by joelfmrodrigues** [comment]
@BaronSparky I used the below as a workaround and worked for me. In summary, create a folder outside of the Function (under site root folder) and download the file from SharePoint. The folder for templates is located outside of the Function itself as we don't have write permissions on the Function folder

$TEMPLATESPATH = "$PSScriptRoot/../../ProvisioningTemplates"
mkdir $TEMPLATESPATH -Force
Get-PnPFile -Url "$SPO_SITE_COLLECTION_RELATIVE_URL/$SPO_SITE_TEMPLATES_LIST_URL/$templateFile" -Path $TEMPLATESPATH -FileName $templateFile -AsFile -Force


![image](https://user-images.githubusercontent.com/19577724/116420940-54d6dd80-a836-11eb-9f83-ad869e669a87.png)

But this is a workaround...would be awesome to be able to reference the template by URL :)

**Reply by BaronSparky** [reply]
Thanks @joelfmrodrigues.

I was using something similar using Get-PnPFile with a blob folder but I had put it in the function directory which meant that I had to remove the "WEBSITE_RUN_FROM_PACKAGE" as deploying from VS Code.

Much prefer the idea of the folder being outside of the function so I can go back to using VS Code again so I'll adopt that approach :D 

Thanks.
